mlir/lib/ExecutionEngine/CMakeLists.txt

```diff
@@ -177,6 +178,39 @@ if(LLVM_ENABLE_PIC)
     target_link_options(mlir_async_runtime PRIVATE "-Wl,-exclude-libs,ALL")
   endif()
 
+  if (MLIR_RUN_ARM_SME_TESTS)
+    if (NOT DEFINED LLVM_MAIN_SRC_DIR)
+      message(FATAL_ERROR "LLVM_MAIN_SRC_DIR must be provided to build the ArmSME runtime.")
+    endif()
+
+    if (NOT DEFINED MLIR_ARM_SME__CAN_ASSEMBLE_ARM_SME)
+      # This should work on an AArch64 host with a recent version of clang.
+      file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/has_arm_sme_check.S
+      ".arch armv9-a+sme
+      .global main
+      .type   main, %function
+      main: smstart
+      .size   main, .-main")
+      try_compile(MLIR_ARM_SME__CAN_ASSEMBLE_ARM_SME ${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_BINARY_DIR}/has_arm_sme_check.S)
+    endif()
+
+    if (NOT MLIR_ARM_SME__CAN_ASSEMBLE_ARM_SME)
+      message(FATAL_ERROR "Host compiler must be able to assemble AArch64 SME instructions to build the ArmSME runtime.")
+    endif()
+
+    # FIXME: This is very far from ideal, but enabling compiler-rt in the main
```

Comment:
I managed to build a minimal (aarch64 builtins only) standalone compiler-rt on aarch64 with the following:
```
# Get recent clang with SME support. This is the most recent but older versions (within reason) should work as well.
wget https://github.com/llvm/llvm-project/releases/download/llvmorg-17.0.4/clang+llvm-17.0.4-aarch64-linux-gnu.tar.xz
tar xvf clang+llvm-17.0.4-aarch64-linux-gnu.tar.xz

cd <path-to-llvm-project>

cmake -G Ninja \
  -S compiler-rt -B build-compiler-rt \
  -DCMAKE_C_COMPILER=<path-to>/clang+llvm-17.0.4-aarch64-linux-gnu/bin/clang-17 \
  -DCMAKE_CXX_COMPILER=<path-to>/clang+llvm-17.0.4-aarch64-linux-gnu/bin/clang++ \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_INSTALL_PREFIX=<install-path> \
  -DLLVM_CMAKE_DIR=cmake \
  -DCOMPILER_RT_DEFAULT_TARGET_ONLY=ON \
  -DCMAKE_C_COMPILER_TARGET=aarch64-linux-gnu \
  -DCOMPILER_RT_BUILD_GWP_ASAN=OFF \
  -DCOMPILER_RT_BUILD_LIBFUZZER=OFF \
  -DCOMPILER_RT_BUILD_MEMPROF=OFF \
  -DCOMPILER_RT_BUILD_ORC=OFF \
  -DCOMPILER_RT_BUILD_PROFILE=OFF \
  -DCOMPILER_RT_BUILD_SANITIZERS=OFF \
  -DCOMPILER_RT_BUILD_XRAY=OFF

ninja -C build-compiler-rt install

tree <install-path>
install
└── lib
    └── linux
        ├── clang_rt.crtbegin-aarch64.o
        ├── clang_rt.crtend-aarch64.o
        └── libclang_rt.builtins-aarch64.a
```

As far as I can tell there's three approaches here:

1. Build compiler-rt together with llvm+clang+mlir.
2. Build standalone compiler-rt separately (as above).
3. Build minimal SME runtime from sources when building llvm+mlir (as you're doing here).

All options require recent host toolchain with SME support.

Options 1 and 3 the runtime is integrated into normal build. I guess for both these options it wouldn't be necessary to add a CMake flag to specify the path to the runtime. Option 2 the runtime is built separately and a CMake flag would be necessary.

Options 1 and 3 may require updating existing build to update host toolchain if current doesn't support SME. Option 2 existing builds wouldn't need to change, except for adding the CMake flag with path to runtime lib.

Option 1 requires building Clang. I like this option the least, it requires building much more than we need as you say.

Comparing options 2 and 3. Option 2 requires fewer changes to MLIR, most (all?) of the changes in this CMake file could be removed? Option 2 would also be insulated from changes to the SME runtime implementation in compiler-rt if it were to ever change.

Food for thought. I'm gravitating towards Option 2, let us know if I've missed anything or what you think.

Reply:
Note that `compiler-rt` only builds the builtin as a static library, so we can't load those at runtime in the MLIR tests. 
```
      add_compiler_rt_runtime(clang_rt.builtins
                              STATIC
                              ARCHS ${arch}
                              DEPS ${deps_${arch}}
                              SOURCES ${${arch}_SOURCES}
                              DEFS ${BUILTIN_DEFS}
                              CFLAGS ${BUILTIN_CFLAGS_${arch}}
                              PARENT_TARGET builtins)
```

Reply:
So, I think option 2 likely requires changing compiler-rt to allow building the builtins into a shared library (probably just the SME builtins).  This would be via some flag on the build like `COMPILER_RT_BUILD_STANDALONE_AARCH64_SME_RUNTIME`.

Reply:
I'm not sure how option 1 would work, since by default `compiler-rt` won't build a shared library we can use, and the `LLVM_ENABLE_RUNTIMES="compiler-rt"` does not seem to expose options from compiler-rt into the main build.

Reply:
> Note that `compiler-rt` only builds the builtins as a static library, so we can't load those at runtime in the MLIR tests.

Ah, I didn't consider that. Surprised that's not configurable, I wonder if there's a reason for only building as static.

Reply:
I think these libraries are intended to be linked into your binary by clang, so in that case it makes sense for them to be static. There is an option to build parts into a shared library, for example:
```
option(COMPILER_RT_BUILD_STANDALONE_LIBATOMIC
  "Build standalone shared atomic library."
  OFF)
```
Which is what gave me the idea for the `COMPILER_RT_BUILD_STANDALONE_AARCH64_SME_RUNTIME` option.

Reply:
> Options 1 and 3 may require updating existing build to update host toolchain if current doesn't support SME. 

IIUC, all options will require a "recent" toolchain, right?

My view is that:
* Option 1 would be the cleanest/simplest and would work fine for pre-commit/post-commit testing (as Buildbots tend to build everything anyway). However, this would be too painful for developers.
* Option 2 would force us to use 2-stage builds for testing. This would be more work for developers and potentially extra set-up for buildbots.
* Options 3 is the simplest from the developer's perspective. And that's what I'd personally optimise for. I do wonder though - wouldn't it be possible to add this target in `compiler-rt` instead?

Reply:
> > Options 1 and 3 may require updating existing build to update host toolchain if current doesn't support SME.
> 
> IIUC, all options will require a "recent" toolchain, right?

Yeah, but for option 2 that toolchain wouldn't need to be used to build llvm+mlir, only to build compiler-rt as standalone.

> My view is that:
> 
>     * Option 1 would be the cleanest/simplest and would work fine for pre-commit/post-commit testing (as Buildbots tend to build everything anyway). However, this would be too painful for developers.

As Ben pointed out, this is currently possible since compiler-rt builds static libs.

Reply:
> Yeah, but for option 2 that toolchain wouldn't need to be used to build llvm+mlir, only to build compiler-rt as standalone.

Ah, thanks! IMHO, this sort of things tend to be tricky to document, i.e. "use whatever toolchain for A, and then something special for B". It can be done, just a bit tricky.

> As Ben pointed out, this is currently possible since compiler-rt builds static libs.

Yes, but we shouldn't assume that that can't be updated, right? (also - did you mean _not_ possible?)

Reply:
>Yes, but we shouldn't assume that that can't be updated, right? (also - did you mean not possible?)

Of course, and yeah I meant isn't possible :)